packages/gatsby/src/redux/plugin-runner.ts

```diff
@@ -41,12 +42,35 @@ interface ICreatePageAction {
 }
 
 export const startPluginRunner = (): void => {
-  emitter.on(`CREATE_PAGE`, (action: ICreatePageAction) => {
-    const page = action.payload
-    apiRunnerNode(
-      `onCreatePage`,
-      { page, traceId: action.traceId, parentSpan: action.parentSpan },
-      { pluginSource: action.plugin.name, activity: action.activity }
-    )
-  })
+  const plugins = store.getState().flattenedPlugins
+  const pluginsImplementingOnCreatePage = plugins.filter(plugin =>
+    plugin.nodeAPIs.includes(`onCreatePage`)
+  )
+  const pluginsImplementingOnCreateNode = plugins.filter(plugin =>
+    plugin.nodeAPIs.includes(`onCreatePage`)
+  )
+  if (pluginsImplementingOnCreatePage.length > 0) {
+    emitter.on(`CREATE_PAGE`, (action: ICreatePageAction) => {
+      const page = action.payload
+      apiRunnerNode(
+        `onCreatePage`,
+        { page, traceId: action.traceId, parentSpan: action.parentSpan },
+        { pluginSource: action.plugin.name, activity: action.activity }
+      )
+    })
+  }
+
+  // We make page nodes outside of the normal action for speed so we manually
+  // call onCreateNode here for SitePage nodes.
+  if (pluginsImplementingOnCreateNode.length > 0) {
+    emitter.on(`CREATE_NODE`, (action: ICreateNodeAction) => {
+      const node = action.payload
+      if (node.internal.type === `SitePage`) {
+        apiRunnerNode(`onCreateNode`, {
+          node,
+          traceTags: { nodeId: node.id, nodeType: node.internal.type },
+        })
+      }
+    })
+  }
```

Comment:
should we maybe just trigger `apiRunner` inside `createPage` action creator? We already do so for `onCreateNode` in https://github.com/gatsbyjs/gatsby/blob/3bcb6e26d5a44ee0687894eb4b02dcfcf013770a/packages/gatsby/src/redux/actions/public.js#L802-L821 after `createNode`. Less emitters the better IMO (easier to follow code)

Reply:
hmm yeah good idea — also extra function calls really do start to add up at scale.

Reply:
same with `onCreatePage` too then I suppose

Reply:
welp, seems like we need to use thunk to use similar setup and that actually make things less performant so let's leave it as-is